File: Services/analyse/src/main/java/org/example/analyse/Models/enums/BloodType.java
package org.example.analyse.Models.enums;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum BloodType {
    A_POSITIVE(1, "A positive"),
    A_NEGATIVE(2, "A negative"),
    B_POSITIVE(3, "B positive"),
    B_NEGATIVE(4, "B negative"),
    AB_POSITIVE(5, "AB positive"),
    AB_NEGATIVE(6, "AB negative"),
    O_POSITIVE(7, "O positive"),
    O_NEGATIVE(8, "O negative");

    private final int id;
    private final String name;
    private static final Map<Integer, BloodType> BY_ID = new HashMap<>();

    static {
        for (BloodType type : BloodType.values()) {
            BY_ID.put(type.id, type);
        }
    }

    BloodType(int id, String name) {
        this.id = id;
        this.name = name;
    }

    @JsonValue
    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    @JsonCreator
    public static BloodType fromId(int id) {
        BloodType type = BY_ID.get(id);
        if (type == null) {
            throw new IllegalArgumentException("No BloodType with id " + id + " found");
        }
        return type;
    }
}

File: front/src/app/Models/BloodType.ts
export const BloodType = {
  A_POSITIVE: { id: 1, name: 'A positive' },
  A_NEGATIVE: { id: 2, name: 'A negative' },
  B_POSITIVE: { id: 3, name: 'B positive' },
  B_NEGATIVE: { id: 4, name: 'B negative' },
  AB_POSITIVE: { id: 5, name: 'AB positive' },
  AB_NEGATIVE: { id: 6, name: 'AB negative' },
  O_POSITIVE: { id: 7, name: 'O positive' },
  O_NEGATIVE: { id: 8, name: 'O negative' }
} as const;

// Utility to get BloodType by ID
export function getBloodTypeById(id: number): typeof BloodType[keyof typeof BloodType] | undefined {
  return Object.values(BloodType).find(type => type.id === id);
}